mesonbuild/wrap/wrap.py

```diff
@@ -129,35 +170,77 @@ def get(self, key: str) -> str:
-    def __init__(self, subdir_root: str, wrap_mode=WrapMode.default, current_subproject: str = ''):
+    def __init__(self, subdir_root: str, wrap_mode=WrapMode.default):
         self.wrap_mode = wrap_mode
         self.subdir_root = subdir_root
-        self.current_subproject = current_subproject
         self.cachedir = os.path.join(self.subdir_root, 'packagecache')
         self.filesdir = os.path.join(self.subdir_root, 'packagefiles')
+        self.wraps = {} # type: T.Dict[str, PackageDefinition]
+        self.provided_deps = {} # type: T.Dict[str, PackageDefinition]
+        self.provided_programs = {} # type: T.Dict[str, PackageDefinition]
+        self.load_wraps()
 
-    def resolve(self, packagename: str, method: str) -> str:
+    def load_wraps(self):
+        if not os.path.isdir(self.subdir_root):
+            return
+        for f in os.listdir(self.subdir_root):
+            fname = os.path.join(self.subdir_root, f)
+            # Ignore not .wrap files, and reserved directories.
+            if (os.path.isfile(fname) and not fname.endswith('.wrap')) or \
+               f in ['packagecache', 'packagefiles']:
+                continue
+            wrap = PackageDefinition(fname)
+            # We could have added a dummy package definition for the directory,
+            # replace it now with the proper wrap. This happens if we already
+            # downloaded the subproject into 'foo-1.0' directory and we now found
+            # 'foo.wrap' file.
+            if wrap.directory in self.wraps:
+                del self.wraps[wrap.directory]
+            self.wraps[wrap.name] = wrap
+            for k in wrap.provided_deps.keys():
+                self.provided_deps[k] = wrap
+            for k in wrap.provided_programs:
+                self.provided_programs[k] = wrap
+
+    def find_dep_provider(self, packagename: str):
+        # Return value is in the same format as fallback kwarg:
+        # ['subproject_name', 'variable_name'], or 'subproject_name'.
+        wrap = self.provided_deps.get(packagename)
+        if wrap:
+            dep_var = wrap.provided_deps.get(packagename)
+            if dep_var:
+                return [wrap.name, dep_var]
+            return wrap.name
+        return None
+
+    def find_program_provider(self, names: T.List[str]):
+        for name in names:
+            wrap = self.provided_programs.get(name)
+            if wrap:
+                return wrap.name
+        return None
+
+    def get_provided_deps(self, packagename: str):
+        wrap = self.wraps.get(packagename)
+        return wrap.provided_deps if wrap else {}
+
+    def resolve(self, packagename: str, method: str, current_subproject: str = '') -> str:
+        self.current_subproject = current_subproject
         self.packagename = packagename
-        self.wrap, self.directory = get_directory(self.subdir_root, self.packagename)
+        self.directory = packagename
+        self.wrap = self.wraps.get(packagename)
```

Comment:
I agree that needs to be done in a follow-up PR.

Reply:
Oh, actually that made unit test fail because it could raise WrapNotFoundException and in exception handling in interpreter it uses r.directory. I think that's part of the refactoring needed to stop setting context on self in resolve() method.